File: src/System.Commands/Logging/ConsoleLoggerScope.vb
Friend Class ConsoleLoggerScope
    Implements IDisposable

    Private disposedValue As Boolean
    Private sources As IList(Of ConsoleLoggerScope)()

    ''' <summary>
    ''' Gets or sets the state to log.
    ''' </summary>
    Public ReadOnly Property State As Object

    ''' <summary>
    ''' Initializes a new instance of the <see cref="ConsoleLoggerScope"/> class.
    ''' </summary>
    Sub New(source As IList(Of ConsoleLoggerScope)(), state As Object)
        ' Initialize any resources if needed
        Me.sources = source
        Me.State = state
    End Sub

    ''' <summary>
    ''' Releases unmanaged and - optionally - managed resources.
    ''' </summary>
    ''' <param name="disposing"><c>true</c> to release both managed and unmanaged resources; <c>false</c> to release only unmanaged resources.</param>
    Protected Overridable Sub Dispose(disposing As Boolean)
        If Not disposedValue Then
            If disposing Then
                ' TODO: dispose managed state (managed objects)
            End If

            ' TODO: free unmanaged resources (unmanaged objects) and override finalizer
            ' TODO: set large fields to null
            disposedValue = True
            TryCast(sources, IList(Of ConsoleLoggerScope)).Remove(Me)
        End If
    End Sub

    ' ' TODO: override finalizer only if 'Dispose(disposing As Boolean)' has code to free unmanaged resources
    ' Protected Overrides Sub Finalize()
    '     ' Do not change this code. Put cleanup code in 'Dispose(disposing As Boolean)' method
    '     Dispose(disposing:=False)
    '     MyBase.Finalize()
    ' End Sub

    ''' <summary>
    ''' Performs application-defined tasks associated with freeing, releasing, or resetting unmanaged resources.
    ''' </summary>
    Public Sub Dispose() Implements IDisposable.Dispose
        ' Do not change this code. Put cleanup code in 'Dispose(disposing As Boolean)' method
        Dispose(disposing:=True)
        GC.SuppressFinalize(Me)
    End Sub
End Class

File: src/Commands/Logging/ConsoleLoggerScope.cs
namespace Commands.Logging
{
	/// <summary>
	/// Represents a logical logging scope for <see cref="ConsoleLogger"/>.
	/// </summary>
	/// <remarks>
	/// Instances of this type are returned from <see cref="ILogger.BeginScope{TState}(TState)"/> and will invoke
	/// the provided cleanup action when disposed to remove the scope from the logger's internal stack.
	/// </remarks>
	internal sealed class ConsoleLoggerScope : IDisposable
	{
		private bool disposedValue;
		private readonly Action? _OnDisposing;

		/// <summary>
		/// Gets the state object associated with this scope.
		/// </summary>
		public object State { get; private set; }

		/// <summary>
		/// Initializes a new instance of the <see cref="ConsoleLoggerScope"/> class.
		/// </summary>
		/// <param name="state">The state object associated with the scope. This value is stored and later written by the logger when logging messages within the scope.</param>
		/// <param name="onDisposing">An action invoked when the scope is disposed. Typically used to remove the scope from the logger's internal collection.</param>
		public ConsoleLoggerScope(object state, Action onDisposing)
		{
			State = state;
			_OnDisposing = onDisposing;
		}

		/// <summary>
		/// Releases resources used by the instance.
		/// </summary>
		/// <param name="disposing">
		/// <c>true</c> when called from <see cref="Dispose"/> to indicate managed resources should be released;
		/// <c>false</c> when called from a finalizer.
		/// </param>
		private void Dispose(bool disposing)
		{
			if (!disposedValue)
			{
				if (disposing)
				{
					// TODO: dispose managed state (managed objects)
				}

				// TODO: free unmanaged resources (unmanaged objects) and override finalizer
				// TODO: set large fields to null
				disposedValue = true;
				_OnDisposing?.Invoke();
			}
		}

		// // TODO: override finalizer only if 'Dispose(bool disposing)' has code to free unmanaged resources
		// ~ConsoleLoggerScope()
		// {
		//     // Do not change this code. Put cleanup code in 'Dispose(bool disposing)' method
		//     Dispose(disposing: false);
		// }

		/// <summary>
		/// Performs application-defined tasks associated with freeing, releasing, or resetting unmanaged resources.
		/// Invokes the cleanup action provided at construction time.
		/// </summary>
		public void Dispose()
		{
			// Do not change this code. Put cleanup code in 'Dispose(bool disposing)' method
			Dispose(disposing: true);
			GC.SuppressFinalize(this);
		}
	}

	/// <summary>
	/// Helper container used by <see cref="ConsoleLogger"/> to hold a reference to a <see cref="ConsoleLoggerScope"/>.
	/// </summary>
	internal sealed class ConsoleLoggerScopeItem
	{
		/// <summary>
		/// Gets or sets the associated <see cref="ConsoleLoggerScope"/>.
		/// </summary>
		public ConsoleLoggerScope? Scope { get; set; }
	}
}